### ISSUE TITLE
Quality array not set correctly on `SpectrumDatasetOnOff.counts_off`

### ISSUE THREAD

**Issue by adonath** [gammapy/gammapy#2208]
@lmohrmann reported a rather severe bug, that I introduced in https://github.com/gammapy/gammapy/commit/b05084654a634b48825cc64d4ff787f89f5c61c4#diff-40398203031fc25ce19075b9d61b3ceaR258. In the `.mask_safe` setter the quality array is not set on the `.counts_off`  data member. For the dataset and likelihood evaluation this does not matter, but the `SpectrumDatasetOnOffStacker` relies on the correct setting of this mask. The bug went un-noticed, because in the testing we still use the old API via the `.lo_threshold` and `hi_threshold` setters, which is unaffected.  

 There are two possible easy fixes:

1. Set `counts_off.quality` in the `mask_safe` setter.
2. Change the `SpectrumDatasetOnOffStacker` to only use `mask_safe` and not `.quality`

For v0.13 we should definitely continue the clean-up process and remove the `quality` handling in `gammapy.spectrum` completely. I.e. only rely on the `mask_safe` and not duplicate the information on the `counts` and `counts_off` data attributes. I think it was a previous design mistake in `gammapy.spectrum` to build our internal data model after the OGIP serialisation. 

Is this a first occasion for a `v0.12.1` release, @deil and @registerrier?

**Comment by cdeil** [comment]
Sounds related to the remaining possibly errors that pylint points out for Gammapy:
```
$ make pylint
pylint -E gammapy/ \
    --ignore=_astropy_init.py,gammapy/extern \
    -d E0611,E1101,E1103 \
    --msg-template='{C}: {path}:{line}:{column}: {msg} ({symbol})'
************* Module gammapy.maps.hpx
E: gammapy/maps/hpx.py:301:8: 'mult_val' does not support item assignment (unsupported-assignment-operation)
E: gammapy/maps/hpx.py:398:4: 'order' does not support item assignment (unsupported-assignment-operation)
************* Module gammapy.maps.wcsnd
E: gammapy/maps/wcsnd.py:348:12: 'pad_msk' does not support item assignment (unsupported-assignment-operation)
E: gammapy/maps/wcsnd.py:349:12: 'pad_msk' does not support item assignment (unsupported-assignment-operation)
************* Module gammapy.utils.interpolation
E: gammapy/utils/interpolation.py:123:37: Access to member '_unit' before its definition line 126 (access-member-before-definition)
************* Module gammapy.utils.energy
E: gammapy/utils/energy.py:46:15: Argument 'dtype' passed by position and keyword in classmethod call (redundant-keyword-arg)
************* Module gammapy.utils.serialization.xml
E: gammapy/utils/serialization/xml.py:330:37: Undefined variable 'parname_registry' (undefined-variable)
************* Module gammapy.utils.serialization.tests.test_xml
E: gammapy/utils/serialization/tests/test_xml.py:150:4: Assigning result of a function call, where the function has no return (assignment-from-no-return)
E: gammapy/utils/serialization/tests/test_xml.py:150:10: No value for argument 'filename' in method call (no-value-for-parameter)
E: gammapy/utils/serialization/tests/test_xml.py:168:4: Assigning result of a function call, where the function has no return (assignment-from-no-return)
E: gammapy/utils/serialization/tests/test_xml.py:168:10: No value for argument 'filename' in method call (no-value-for-parameter)
************* Module gammapy.spectrum.core
E: gammapy/spectrum/core.py:399:8: 'self.quality' does not support item assignment (unsupported-assignment-operation)
E: gammapy/spectrum/core.py:412:8: 'self.quality' does not support item assignment (unsupported-assignment-operation)
************* Module gammapy.spectrum.tests.test_dataset
E: gammapy/spectrum/tests/test_dataset.py:477:12: 'npred_summed' does not support item assignment (unsupported-assignment-operation)
************* Module gammapy.astro.population.simulate
E: gammapy/astro/population/simulate.py:67:33: bad operand type for unary -: str (invalid-unary-operand-type)
E: gammapy/astro/population/simulate.py:70:33: bad operand type for unary -: str (invalid-unary-operand-type)
E: gammapy/astro/population/simulate.py:71:33: bad operand type for unary -: str (invalid-unary-operand-type)
E: gammapy/astro/population/simulate.py:74:33: bad operand type for unary -: str (invalid-unary-operand-type)
E: gammapy/astro/population/simulate.py:75:33: bad operand type for unary -: str (invalid-unary-operand-type)
E: gammapy/astro/population/simulate.py:76:33: bad operand type for unary -: str (invalid-unary-operand-type)
make: *** [pylint] Error 2
```
We should fix them all, or if invalid, file a bug report with pylint and silence them for now.

> Is this a first occasion for a v0.12.1 release, @deil and @registerrier?

I would prefer we just do v0.13, v0.14, ...

The problem with doing v0.12.1 is that already after v0.12 we had a lot of commits in master, and we are using git tags, not branches currently. So we are not set up yet for the workflow to have bugfix branches with backported important fixes.

Of course we could just fix this in master and put the next git tag and release at v0.12.1. That's an option. The problem there is only that the changelog and what's new for Gammapy becomes more complicated. Another option could be that we just do a v0.13 a bit early, e.g. July 1.

@lmohrmann - if you can wait until July to submit the HESS validation paper, I would suggest we fix this now and use Gammapy master for now, and the do v0.13 in (possibly early July).

@adonath - Can you please create a separate issue to discuss v0.13 or just set a milestone date and put it on the agenda for the call this Friday? Mixing this discussion with the actual bug will be confusing.

**Comment by registerrier** [comment]
I think the two changes need to be made. 
As long as there is a `.quality` property on the `PHACountsSpectrum` we need to maintain its synchronization with the `.safe_mask`. There might be other cases where we would have a problem.

**Comment by cdeil** [comment]
IMO the strategy in Gammapy should be (everywhere, but especially here): remove as much fluff as possible, improve the quality and tests of what remains, and then (or in parallel) refactor towards a better design / API, always keeping established use cases working and tests passing in each PR. A series of 100 small PRs that go in very quickly.

**Comment by lmohrmann** [comment]
@cdeil I don't necessarily need this fix or a new release, there's an easy workaround that I can use (essentially copying the `quality` attribute from `counts` to `counts_off`). So we'll stay with v0.12 and use that for the paper.
But users that are not aware of this bug will get wrong results in the 1D spectral analysis, and there's no warning or error making them aware of this...

**Comment by adonath** [comment]
This was fixed in #2231.
